### PR TITLE
Add ShemarooMeIE.

### PR DESCRIPTION
**Pull request by Ashish0804**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

closes https://github.com/yt-dlp/yt-dlp/issues/307
This is a bare bones implementation and can be improved a lot. I'll try to add more features to this in future. 

Also the release date and description searching takes a lot of time (well it's noticeable), i am open to ideas to make it much faster.

### CODE REVIEW

File: yt_dlp/extractor/shemaroome.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        params_pattern = r'params_for_player\ =\ \"(?P<data>[^|]+)\|key=(?P<key>[^|]+)\|image=(?P<thumbnail>[^|]+)\|title=(?P<title>[^|]+)'
+        m = re.search(params_pattern, webpage)
+        data = bytes_to_intlist(compat_b64decode(m.group('data')))
+        key = bytes_to_intlist(compat_b64decode(m.group('key')))
+        iv = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
Comment (pukkandan) [inline yt_dlp/extractor/shemaroome.py:39]:
```suggestion
        iv = [0] * 16
```

File: yt_dlp/extractor/shemaroome.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        m3u8_url = intlist_to_bytes(aes_cbc_decrypt(data, key, iv))
+        m3u8_url = m3u8_url[:-compat_ord((m3u8_url[-1]))].decode('ascii')
+        formats = self._extract_m3u8_formats(m3u8_url, video_id, fatal=False)
+        self._sort_formats(formats)
+        release_date = self._html_search_regex(r'\<span\ itemprop\=\"uploadDate\"\>(?P<uploaddate>\S+)\<\/span\>',
+                                               webpage, 'release_date', default="NA", fatal=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/shemaroome.py:45]:
```suggestion
                                               webpage, 'release_date', fatal=False)
```

File: yt_dlp/extractor/shemaroome.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        description = self._html_search_regex(r'\<p\ class\=\"float-left\ w-100\ app-color1\ font-regular\"\>(?P<description>[^\<]+)\<\/p\>',
+                                              webpage, 'description', default="NA", fatal=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/shemaroome.py:47]:
```suggestion
                                              webpage, 'description', fatal=False)
```
default should be `None`

File: yt_dlp/extractor/shemaroome.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    _VALID_URL = r'(?:https?://)(?:www\.)?shemaroome\.com/(?:movies|shows)/(?P<id>\S+)'
+    _TESTS = [{
+        'url': 'https://www.shemaroome.com/movies/dil-hai-tumhaara',
+        'info_dict': {
+            'id': 'dil-hai-tumhaara',
+            'ext': 'mp4',
+            'title': 'Dil Hai Tumhaara',
+            'release_date': '2002-09-06',
+            'thumbnail': 'https://daex9l847wg3n.cloudfront.net/shemoutputimages/Dil-Hai-Tumhaara/60599346a609d2faa3000020/large_16_9_1616436538.jpg?1616483693',
+            'description': "A chirpy young girl Shalu craves for the love of her mother Sarita, but Sarita is partial towards her elder daughter Nimmi. Shalu is in love with her employer's son, Dev, but when Sarita proposes Nimmi's marriage to Dev, Shalu is forced to choose between her mother's acceptance and Dev's love. Matters worsen when Shalu is told the truth about her heritage.",
```
Comment (pukkandan) [inline yt_dlp/extractor/shemaroome.py:28]: use `md5:`

**Review by pukkandan — COMMENTED** [review]
I'll have a look at how to improve the regex later

File: yt_dlp/extractor/shemaroome.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        params_pattern = r'params_for_player\ =\ \"(?P<data>[^|]+)\|key=(?P<key>[^|]+)\|image=(?P<thumbnail>[^|]+)\|title=(?P<title>[^|]+)'
+        m = re.search(params_pattern, webpage)
+        data = bytes_to_intlist(compat_b64decode(m.group('data')))
+        key = bytes_to_intlist(compat_b64decode(m.group('key')))
+        iv = [0] * 16
+        m3u8_url = intlist_to_bytes(aes_cbc_decrypt(data, key, iv))
+        m3u8_url = m3u8_url[:-compat_ord((m3u8_url[-1]))].decode('ascii')
+        formats = self._extract_m3u8_formats(m3u8_url, video_id, fatal=False)
+        self._sort_formats(formats)
+        release_date = self._html_search_regex(r'\<span\ itemprop\=\"uploadDate\"\>(?P<uploaddate>\S+)\<\/span\>',
```
Comment (nao20010128nao) [inline yt_dlp/extractor/shemaroome.py:44]: May not improve regex match speed, but this is more shorter: `<span\sitemprop="uploadDate">(?P<uploaddate>\S+)</span>`

File: yt_dlp/extractor/shemaroome.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                                               webpage, 'release_date', fatal=False)
+        description = self._html_search_regex(r'\<p\ class\=\"float-left\ w-100\ app-color1\ font-regular\"\>(?P<description>[^\<]+)\<\/p\>',
+                                              webpage, 'description', fatal=False)
```
Comment (nao20010128nao) [inline yt_dlp/extractor/shemaroome.py:47]: here too: `<p\sclass="float-left\sw-100\sapp-color1\sfont-regular">(?P<description>[^<]+)</p>`

File: yt_dlp/extractor/shemaroome.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+from ..utils import (
+    bytes_to_intlist,
+    intlist_to_bytes,
+    unified_strdate,
+    url_or_none,
+)
+
+
+class ShemarooMeIE(InfoExtractor):
+    _VALID_URL = r'(?:https?://)(?:www\.)?shemaroome\.com/(?:movies|shows)/(?P<id>\S+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/shemaroome.py:21]:
```suggestion
    _VALID_URL = r'(?:https?://)(?:www\.)?shemaroome\.com/(?:movies|shows)/(?P<id>[-\w]+)'
```
or
```suggestion
    _VALID_URL = r'(?:https?://)(?:www\.)?shemaroome\.com/(?:movies|shows)/(?P<id>[^/?#]+)'
```

**Pull request merged**